File: python/trees.py
#! /usr/bin/env pyth

from Queue import Queue
from copy import deepcopy

class Node:
    def __init__(self, value):
        self.left  = None
        self.right = None
        self.key   = value
        
class Tree:
    def __init__(self):
        self.root = None
    
    def __addNode__(self, value, leaf):
        if value <= leaf.key:
            if leaf.left is not None:
                self.__addNode__(value, leaf.left)
            else:
                leaf.left = Node(value)                
        else:
            if leaf.right is not None:
                self.__addNode__(value, leaf.right)
            else:
                leaf.right = Node(value)
           
    def addNode(self, value):
        if self.root is None:
            self.root = Node(value)
        else:
            self.__addNode__(value, self.root)

    def inOrder(self, node):
        if node is not None:
            self.inOrder(node.left)
            nodeAction(node)
            self.inOrder(node.right)
        
    def preOrder(self, node):
        if node is not None:
            nodeAction(node)
            self.preOrder(node.left)
            self.preOrder(node.right)
        
    def postOrder(self, node):
        if node is not None:
            self.postOrder(node.left)
            self.postOrder(node.right)
            nodeAction(node)
            
    def lookUp(self, value, node=None, parent=None):    
        if value < node.key:
            if node.left is None:
                return None, node
            else:
                return self.lookUp(value, node.left, node)
        elif value > node.key:
            if node.right is None:
                return None, node
            else:
                return self.lookUp(value, node.right, node)                  
        return node, parent
        
    def getDepth(self, node):
        if node is None:
            return 0
        else:
            leftDepth = self.getDepth(node.left)
            rightDepth = self.getDepth(node.right)
            if leftDepth > rightDepth:
                return leftDepth + 1
            else:
                return rightDepth + 1
                
    def __findMin__(self, node):
        while node.left is not None:
            node = node.left
        return node
    
    def __findMax__(self, node):
        while node.right is not None:
            node = node.right
        return node
                
    def __deleteNode__(self, value, curNode, parent=None):
        node, parent = self.lookUp(value, curNode, parent)
        if node is not None:
            if node.left is None and node.right is None:
                if parent.left.key == value: parent.left = None
                elif parent.right.key == value: parent.right = None
                del node
            elif node.left is None and node.right is not None:
                if parent.left.key == value: 
                    parent.left = node.right
                    del node                    
                elif parent.right.key == value: 
                    parent.right = node.right
                    del node
            elif node.left is not None and node.right is None:
                if parent.right.key == value: 
                    parent.right = node.left
                    del node                    
                elif parent.left.key == value: 
                    parent.left = node.right
                    del node
            else: # target node has children on either side
                l = self.getDepth(node.left)
                r = self.getDepth(node.right)
                tmp = None
                if r > l:                   
                    tmp = self.__findMin__(node.right)
                    node.key = tmp.key   
                    self.__deleteNode__(node.key, node.right, node)
                else:
                    tmp = self.__findMax__(node.left)
                    node.key = tmp.key
                    self.__deleteNode__(node.key, node.left, node)    
            
    def deleteNode(self, value):
        self.__deleteNode__(value, self.root)         
                                                      
        
def nodeAction(node):
    print str(node.key) + " ",        
        
def depthTraversal(tree):
    tree.preOrder(tree.root)
        
def breadthTraversal(tree):
    toTraverse = Queue(maxsize = 0)
    toTraverse.put(tree.root)
    while not toTraverse.empty():
        currentNode = traversal.get()
        nodeAction(currentNode)
        if currentNode.left  is not None: 
            toTraverse.put(currentNode.left)
        if currentNode.right is not None: 
            toTraverse.put(currentNode.right)



def main():
    tree = Tree()
    tree.addNode(8)
    tree.addNode(3)
    tree.addNode(10)
    tree.addNode(1)
    tree.addNode(6)
    tree.addNode(4)
    tree.addNode(7)
    tree.addNode(14)
    tree.addNode(13)
    tree.addNode(15)
    tree.addNode(9)

    '''
    Our tree should be:
            8
        /       \    
       3        10
      / \      /  \ 
     1   6    9    14
        / \        / \
        4 7      13  15 
    '''
    
    # output should be: 1 3 4 6 7 8 9 10 13 14 15
    print "In-order traversal:"
    tree.inOrder(tree.root) 
    
    # output should be: 8 3 1 6 4 7 10 9 14 13 15
    print "\nPre-order traversal:"
    tree.preOrder(tree.root)
    
    # output should be:  1 4 7 6 3 9 13 15 14 10 8
    print "\nPost-order traversal:"
    tree.postOrder(tree.root)
    
    print "\nDepth-first traversal:"
    depthTraversal(tree)
    
    # output should be: 8 3 10 1 6 9 14 4 7 13 15
    print "\nBreadth-first traversal:"
    breadthTraversal(tree)    
    
    #print "\nLookup:"
    #x = tree.lookUp(32)
    #if x[0] is not None: print "x[0]: " + str(x[0].key)
    #if x[1] is not None: print "x[1]: " + str(x[1].key)
    
    tree2 = deepcopy(tree)
    x = 1 
    tree2.deleteNode(x)
    print "\nBreadth-first traversal after deletion:" + str(x)
    breadthTraversal(tree2)  

    x = 3
    tree2.deleteNode(x)    
    print "\nBreadth-first traversal after deletion:" + str(x)
    breadthTraversal(tree2)   
    
    x = 15
    tree2.deleteNode(x)
    print "\nBreadth-first traversal after deletion:" + str(x)
    breadthTraversal(tree2)  
    x = 14  
    tree2.deleteNode(x)        
    print "\nBreadth-first traversal after deletion:" + str(x)
    breadthTraversal(tree2)       
  
    print "\nNew tree!!!"
    x = 14
    tree3 = deepcopy(tree)
    tree3.deleteNode(x)   
    print "\nBreadth-first traversal after deletion:" + str(x)
    breadthTraversal(tree3)
    
    print "\nDepth-first"
    depthTraversal(tree3)

    
if __name__ == "__main__":
    main()

File: Cplusplus/trees.cpp
#include <iostream>

#define DEBUG

using namespace std;

class Node {
    int key;
    Node* left;
    Node* right;
    
public:
    Node() {
        key   = -1;
        left  = NULL;
        right = NULL;
    };
    
    void setKey(int value)      { key   = value; };
    void setLeft(Node *aLeft)   { left  = aLeft; };
    void setRight(Node *aRight) { right = aRight; };
    int Key()     { return key; };
    Node* Left()  { return left; };
    Node* Right() { return right; };
};

class Tree {
    Node* root;
public:
    Tree();
    ~Tree();
    Node *Root() { return root; };
    void addNode(int key);
    void inOrder(Node* n);
    void preOrder(Node* n);
    void postOrder(Node* n);
private:
    void addNode(int key, Node* leaf);
    void freeNode(Node* leaf);
    void printKey(Node* n);
};

Tree::Tree() { root = NULL; }
Tree::~Tree() { freeNode(root); }

void Tree::freeNode(Node* leaf) {
    if (leaf != NULL) {
        freeNode(leaf->Left());
        freeNode(leaf->Right());
        delete leaf;
    }
}

// public
void Tree::addNode(int key) {
    if (root == NULL) {
        #ifdef DEBUG
        cout << "add root node ... " << key << endl;
        #endif
        Node* n = new Node();
        n->setKey(key);
        root = n;
    } else {
        #ifdef DEBUG
        cout << "add other node ... " << key << endl;
        #endif
        addNode(key, root);
    }
}

// private
void Tree::addNode(int key, Node *leaf) {
    if (key <= leaf->Key()) {
        if (leaf->Left() != NULL)
            addNode(key, leaf->Left());
        else {
            Node* n = new Node();
            n->setKey(key);
            leaf->setLeft(n);
        }    
    } else {
        if (leaf->Right() != NULL)
            addNode(key, leaf->Right());
        else {
            Node* n = new Node();
            n->setKey(key);
            leaf->setRight(n);   
        }
    }
}

void Tree::printKey(Node* n) {
    #ifdef DEBUG
    cout << n->Key() << " ";
    #endif
}

void Tree::inOrder(Node* n) {
    if (n) {
        inOrder(n->Left());
        printKey(n);     
        inOrder(n->Right());
    }
}

void Tree::preOrder(Node* n) {
    if (n) {
        printKey(n);  
        preOrder(n->Left());   
        preOrder(n->Right());
    }
}

void Tree::postOrder(Node* n) {
    if (n) { 
        postOrder(n->Left());   
        postOrder(n->Right());
        printKey(n); 
    }
}

int main() {
    Tree* tree = new Tree();
    tree->addNode(8);
    tree->addNode(3);
    tree->addNode(10);
    tree->addNode(1);
    tree->addNode(6);
    tree->addNode(4);
    tree->addNode(7);
    tree->addNode(14);
    tree->addNode(13);
    tree->addNode(15);
    tree->addNode(9);
    
    /******************
    
    Our tree should be:
            8
         /    \    
       3        10
      / \      /  \ 
     1   6    9    14
        / \        / \
        4 7      13  15  
    ***********************/
    
    // output should be: 1 3 4 6 7 8 9 10 13 14 15
    cout << "In-order traversal" << endl;
    tree->inOrder(tree->Root());
    cout << endl;
    
    // output should be: 8 3 1 6 4 7 10 9 14 13 15
    cout << "Pre-order traversal" << endl;
    tree->preOrder(tree->Root());
    cout << endl;
    
    // output should be: 1 47 6 3 9 13 15 14 10 8
    cout << "Post-order traversal" << endl;
    tree->postOrder(tree->Root());
    cout << endl;
    
    delete tree;
    return 0;   
}
